File: cpp/234.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
         if(!head || !head->next) 
            return true;
         ListNode *slow = head;
         ListNode *fast = head;
         while(fast->next && fast->next->next){
             slow = slow->next;
             fast = fast->next->next;
         }
         ListNode *newHead = reverseList(slow->next);
         while(newHead){
             if(newHead->val != head->val){
                 return false;
             }
             newHead = newHead->next;
             head = head->next;
         }
         return true;
    }
    
    ListNode* reverseList(ListNode *head){
        ListNode *next = NULL;
        ListNode *pre = NULL;
        while(head){
            next = head->next;
            head->next = pre;
            pre = head;
            head = next;
        }
        return pre;
    }
    
};

File: golang/234/234.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	left := head
	right := reverse(slow)
	for right != nil {
		if left.Val != right.Val {
			return false
		}
		left = left.Next
		right = right.Next
	}

	return true
}

func reverse(h *ListNode) *ListNode {
	var prev *ListNode
	cur := h
	for cur != nil {
		nxt := cur.Next
		cur.Next = prev
		prev = cur
		cur = nxt
	}

	return prev
}
